docs/design/thee_datahall.md

```diff
@@ -0,0 +1,197 @@
+The current requirement for `three_data_hall` and `three_datacenter` is to have 3 AZs and at least two different failure zones per `data_hall`/`datacenter`.
+
+## General Design Goals
+
+The goal of this design is to support the redundancy modes `three_data_hall` and `three_datacenter` in different deployment models:
+
+* Single Kubernetes cluster across multiple AZs.
+* One Kubernetes cluster per AZ.
+
+The implemenation should be flexible enough to give the user a choice of the used AZ.
+
+## Current Implementation
+
+The current version of the operator doesn't support this redundancy mode.
+The locality is currently a mix of some constants, a configurable zone and two settings at the cluster level [locality configuration](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/controllers/cluster_controller.go#L631-L642).
+
+## Proposed Design
+
+The design is split into the two main parts of the deployment: configure the locality and how to deploy the Pods across the cluster.
+
+### Configure locality
+
+The `three_data_hall` deployment requires the `locality_data_hall` locality to be set and the `three_datacenter` requires the `locality_dcid` to be set.
+There must be at least and at most 3 different values for these localities.
+Additionaly to these localties at least two different `locality_zone` per `data_hall`/`datacenter` is required to fullfill the requirements.
+The `locality_zone` can be configured with [FoundationDBClusterFaultDomain](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterfaultdomain).
+In addition to that the user can configure additional variables that the sidecar will use for substition in the [FoundationDBClusterSpec](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterspec) with the `sidecarVariables` list.
+The current implementation would allow to set `dataHall: $AZ` and in `sidecarVariables` we would list `AZ` to define the `locality_data_hall` based on an environment variable that will be replaced by the sidecar with the actual value.
+Instead of having these different mechanisims spread across different settings I propose to have a new `localities` setting in the `FoundationDBClusterSpec`:
+
+```yaml
+localities:
+- key: "locality_data_hall"
+  value: ""
+  valueFromEnv: ""
+  valueFromNode: "topology.kubernetes.io/zone"
+  topologyKey: ""
+- key: "locality_zone"
+  value: ""
+  valueFromEnv: "MyFancyZone"
+  valueFromNode: ""
+  topologyKey: "kubernetes.io/hostname"
+```
+
+The `topologyKey` will be used for `topologySpreadConstraints` and for `PodAntiAffinity` and only has to be set if `valueFromNode` is empty, otherwise it will default to that value.
+The `FoundationDBClusterFaultDomain` would then be reduced to only have `key`, `zoneCount` and `zoneIndex` since the locality will be configured in `localities`.
```

Comment:
How would we be using the `FoundationDBClusterFaultDomain` type in this new model? Is it possible we could deprecate `zoneCount` and `zoneIndex` in favor of the new structures from the design on multi-DC support in the plugin?

Reply:
The idea would be to deprecated `FoundationDBClusterFaultDomain` I just have to look into the special keys for `foundationdb.org/none` and the multi KC case. The later actually sounds good 👍